Now I have all the information needed. Here's the patch:

```diff
--- a/astropy/io/fits/fitsrec.py
+++ b/astropy/io/fits/fitsrec.py
@@ -1276,7 +1276,7 @@
 
         # Replace exponent separator in floating point numbers
         if 'D' in format:
-            output_field.replace(encode_ascii('E'), encode_ascii('D'))
+            output_field[:] = output_field.replace(b'E', b'D')
 
     def tolist(self):
         # Override .tolist to take care of special case of VLF
```